File: ui-tests/PageObjects/Panes/VariableState/VariableStatePane.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects;

namespace UiTests.PageObjects.Panes.VariableState;

/// <summary>
/// Program state pane holding variables and watch expressions.
/// </summary>
public class VariableStatePane : TabObject
{
    private List<VariableStateRecord> _variables = new();

    public VariableStatePane(IPage page, ILocator root, string tabButtonText)
        : base(page, root, tabButtonText)
    {
    }

    public ILocator WatchExpressionTextBox()
        => Root.Locator("#watch");

    public async Task<IReadOnlyList<VariableStateRecord>> ProgramStateVariablesAsync(bool forceReload = false)
    {
        if (forceReload || _variables.Count == 0)
        {
            var locators = await Root.Locator(".value-expanded").AllAsync();
            _variables = locators.Select(l => new VariableStateRecord(l)).ToList();
        }
        return _variables;
    }
}

File: tsc-ui-tests/page-objects/panes/variable-state/variable-state-pane.ts
import type { Locator, Page } from "@playwright/test";
import { ValueComponentView } from "../../components/value-component-view";

/**
 * Program state pane holding variables and watch expressions.
 *
 * Port of ui-tests/PageObjects/Panes/VariableState/VariableStatePane.cs
 * Absorbs the existing state.ts StatePanel.
 */
export class VariableStatePane {
  readonly page: Page;
  readonly root: Locator;
  readonly tabButtonText: string;
  private variables: ValueComponentView[] = [];

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
  }

  watchExpressionTextBox(): Locator {
    return this.root.locator("#watch");
  }

  async programStateVariables(forceReload = false): Promise<ValueComponentView[]> {
    if (forceReload || this.variables.length === 0) {
      const locators = await this.root.locator(".value-expanded").all();
      this.variables = locators.map((l) => new ValueComponentView(l));
    }
    return this.variables;
  }
}
